### PR TITLE
Coredns GA Blog

### PR DESCRIPTION
**Pull request by johnbelamaric**
The blog for the CoreDNS entry for "5 days of Kubernetes". The date in the front matter is wrong, I wasn't sure how to see the content without setting the date back.

@kbarnard10 @miekg @chrisohaver @sandeeprajan @fturib

### CODE REVIEW

File: content/en/blog/_posts/2018-07-10-coredns-ga.md [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+
+In addition to the features described above, CoreDNS is easily extended. It is possible to build custom versions
+of CoreDNS that include your own features. For example, this ability has been used to extend CoreDNS to do recursive resolution
+with the [unbound plugin](https://https://coredns.io/explugins/unbound), to server records directly from a database with the [pdsql plugin](https://coredns.io/explugins/pdsql), and to allow multiple CoreDNS instances to share a common level 2 cache with the [redisc plugin](https://coredns.io/explugins/redisc).
+
+Many other interesting extensions have been added, which you will find on the [External Plugins](https://coredns.io/explugins/) page of the CoreDNS site. One that is really interesting for Kubernetes and Istio users is the [kubernetai plugin](https://coredns.io/explugins/kubernetai), which allows a single CoreDNS instance to connect to multiple Kubernetes clusters and provide service discovery across all of them.
+
+## What's Next?
+
+CoreDNS is an independant project, and as such is developing many features that are not directly
```
Comment (neolit123) [inline content/en/blog/_posts/2018-07-10-coredns-ga.md]: `s/independant/independent/`

**Comment by johnbelamaric** [comment]
/check-cla

Changed my email address.

File: content/en/blog/_posts/2018-07-10-coredns-ga.md (hunk 1 above)
Comment (johnbelamaric) [reply content/en/blog/_posts/2018-07-10-coredns-ga.md]: fixed and all commits squashed

File: content/en/blog/_posts/2018-07-10-coredns-ga.md [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+
+**Author**: John Belamaric (Infoblox)
+
+## Introduction
+
+In Kubernetes 1.11, [CoreDNS](https://coredns.io) has reached General Availability (GA) for DNS-based service discovery, as an alternative to the `kube-dns` addon. This means that CoreDNS will be offered as an option in upcoming versions of the various installation tools. In fact, the `kubeadm` team chose to make it the default option starting with Kubernetes 1.11.
+
+DNS-based service discovery has been part of Kubernetes for a long time with the `kube-dns` cluster addon. This has generally worked pretty well, but there have been some concerns around the reliability, flexibility and security of the implementation.
+
+CoreDNS is a general-purpose, authoritative DNS server that provides a backwards-compatible, but extensible integration with Kubernetes. It resolves the issues seen with `kube-dns`, and offers a number of unique features that solve a wider variety of use cases.
```
Comment (miekg) [inline content/en/blog/_posts/2018-07-10-coredns-ga.md]: add comma after 'extensible' ?

File: content/en/blog/_posts/2018-07-10-coredns-ga.md [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+primary configuration file that CoreDNS uses for configuration of all of its features, even those that are not
+Kubernetes related.
+
+When upgrading from `kube-dns` to CoreDNS using `kubeadm`, your existing ConfigMap will be used to generate the
+customized Corefile for you, including all of the configuration for stub domains, federation, and upstream nameservers. See [Using CoreDNS for Service Discovery](https://kubernetes.io/docs/tasks/administer-cluster/coredns/) for more details.
+
+
+## Bug fixes and enhancements
+
+There are several open issues with `kube-dns` that are resolved in CoreDNS, including:
```
Comment (miekg) [inline content/en/blog/_posts/2018-07-10-coredns-ga.md]: How are these solved? They just work I do I need to fiddle with the Configmap/Corefile. Can we point to CoreDNS issues/PRs ?

File: content/en/blog/_posts/2018-07-10-coredns-ga.md [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+
+  * [dns#55 - Custom DNS entries for kube-dns](https://github.com/kubernetes/dns/issues/55)
+  * [dns#116 - Only one A record set for headless service with pods having single hostname.](https://github.com/kubernetes/dns/issues/116)
+  * [dns#131 - externalName not using stubDomains settings](https://github.com/kubernetes/dns/issues/131)
+  * [dns#167 - enable skyDNS round robin A/AAAA records](https://github.com/kubernetes/dns/issues/167)
+  * [dns#190 - kube-dns cannot run as non-root user](https://github.com/kubernetes/dns/issues/190)
+  * [dns#232 - fix pod hostname to be podname for dns srv records](https://github.com/kubernetes/dns/issues/232)
+
+Most future bug fixes related to DNS will likely appear in CoreDNS before `kube-dns`.
```
Comment (miekg) [inline content/en/blog/_posts/2018-07-10-coredns-ga.md]: that may be a bold claim

Comment (neolit123) [reply content/en/blog/_posts/2018-07-10-coredns-ga.md]: i agree. too much advertising can become anti-advertising.

File: content/en/blog/_posts/2018-07-10-coredns-ga.md [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+
+In `kube-dns`, you can [modify a ConfigMap](https://kubernetes.io/blog/2017/04/configuring-private-dns-zones-upstream-nameservers-kubernetes/) to change the behavior of your service discovery. This allows the addition of
+features such as serving stub domains, modifying upstream nameservers, and enabling federation.
+
+In CoreDNS, you similarly can modify the ConfigMap for the CoreDNS [Corefile](https://coredns.io/2017/07/23/corefile-explained/) to change how service discovery
+works. This Corefile configuration offers many more options than you will find in `kube-dns`, since it is the
+primary configuration file that CoreDNS uses for configuration of all of its features, even those that are not
+Kubernetes related.
+
+When upgrading from `kube-dns` to CoreDNS using `kubeadm`, your existing ConfigMap will be used to generate the
```
Comment (neolit123) [inline content/en/blog/_posts/2018-07-10-coredns-ga.md]: remove " ` " around kube-dns and kubeadm as these are project names?

File: content/en/blog/_posts/2018-07-10-coredns-ga.md [hunk 6]
```diff
@@ -0,0 +1,179 @@
...
+  * [dns#167 - enable skyDNS round robin A/AAAA records](https://github.com/kubernetes/dns/issues/167)
+  * [dns#190 - kube-dns cannot run as non-root user](https://github.com/kubernetes/dns/issues/190)
+  * [dns#232 - fix pod hostname to be podname for dns srv records](https://github.com/kubernetes/dns/issues/232)
+
+Most future bug fixes related to DNS will likely appear in CoreDNS before `kube-dns`.
+
+## Metrics
+
+The functional behavior of the default CoreDNS configuration is the same as `kube-dns`. However,
+one difference you need to be aware of is that the published metrics are not the same. In `kube-dns`,
```
Comment (neolit123) [inline content/en/blog/_posts/2018-07-10-coredns-ga.md]: > In `kube-dns`

remove " ` " - project name?

File: content/en/blog/_posts/2018-07-10-coredns-ga.md [hunk 7]
```diff
@@ -0,0 +1,179 @@
...
+you get separate metrics for `dnsmasq` and `kube-dns` (skydns). In CoreDNS there is a completely
+different set of metrics, since it is all a single process. You can find more details on these
+metrics on the CoreDNS [Prometheus plugin](https://coredns.io/plugins/metrics/) page.
+
+## Some special features
+
+The standard CoreDNS Kubernetes configuration is designed to be backwards compatible with the prior
+`kube-dns` behavior. But with some configuration changes, CoreDNS can allow you to modify how the
```
Comment (neolit123) [inline content/en/blog/_posts/2018-07-10-coredns-ga.md]: remove " ` " around kube-dns  - project name?

File: content/en/blog/_posts/2018-07-10-coredns-ga.md [hunk 8]
```diff
@@ -0,0 +1,179 @@
...
+
+## Introduction
+
+In Kubernetes 1.11, [CoreDNS](https://coredns.io) has reached General Availability (GA) for DNS-based service discovery, as an alternative to the `kube-dns` addon. This means that CoreDNS will be offered as an option in upcoming versions of the various installation tools. In fact, the `kubeadm` team chose to make it the default option starting with Kubernetes 1.11.
+
+DNS-based service discovery has been part of Kubernetes for a long time with the `kube-dns` cluster addon. This has generally worked pretty well, but there have been some concerns around the reliability, flexibility and security of the implementation.
+
+CoreDNS is a general-purpose, authoritative DNS server that provides a backwards-compatible, but extensible integration with Kubernetes. It resolves the issues seen with `kube-dns`, and offers a number of unique features that solve a wider variety of use cases.
+
+In this article, you will learn about the differences in the implementations of `kube-dns` and CoreDNS, and some of the helpful extensions offered by CoreDNS.
```
Comment (neolit123) [inline content/en/blog/_posts/2018-07-10-coredns-ga.md]: [1]

File: content/en/blog/_posts/2018-07-10-coredns-ga.md [hunk 9]
```diff
@@ -0,0 +1,179 @@
...
+layout: blog
+title: "CoreDNS GA for Kubernetes Cluster DNS"
+date: 2018-07-08
+---
+
+**Author**: John Belamaric (Infoblox)
+
+## Introduction
+
+In Kubernetes 1.11, [CoreDNS](https://coredns.io) has reached General Availability (GA) for DNS-based service discovery, as an alternative to the `kube-dns` addon. This means that CoreDNS will be offered as an option in upcoming versions of the various installation tools. In fact, the `kubeadm` team chose to make it the default option starting with Kubernetes 1.11.
```
Comment (neolit123) [inline content/en/blog/_posts/2018-07-10-coredns-ga.md]: remove the " ` " around names like "kube-dns" and "kubeadm"? 
these are project names and not a pod name, a service name or command.

marking this as [1] but i didn't cover all places.

File: content/en/blog/_posts/2018-07-10-coredns-ga.md (hunk 2 above)
Comment (neolit123) [reply content/en/blog/_posts/2018-07-10-coredns-ga.md]: [1]

File: content/en/blog/_posts/2018-07-10-coredns-ga.md [hunk 10]
```diff
@@ -0,0 +1,179 @@
...
+
+DNS-based service discovery has been part of Kubernetes for a long time with the `kube-dns` cluster addon. This has generally worked pretty well, but there have been some concerns around the reliability, flexibility and security of the implementation.
```
Comment (neolit123) [inline content/en/blog/_posts/2018-07-10-coredns-ga.md]: [1]

File: content/en/blog/_posts/2018-07-10-coredns-ga.md [hunk 11]
```diff
@@ -0,0 +1,179 @@
...
+container watches the Kubernetes API and serves DNS records based on the [Kubernetes DNS specification](https://github.com/kubernetes/dns/blob/master/docs/specification.md), while `dnsmasq` provides caching and stub domain support, and `sidecar` provides metrics and health checks.
+
+This setup leads to a few issues that have been seen over time. For one, since `dnsmasq` is written in C, it is not
+memory-safe. This means it is more prone to certain types of security vulnerabilities, and that has led to the need
+for a security-patch release of Kubernetes in the past. Additionally, because `dnsmasq` handles the stub domains,
+but `kubedns` handles the External Services, you cannot use a stub domain in an external service, which is very
+limiting to that functionality (see [dns#131](https://github.com/kubernetes/dns/issues/131)).
+
+All of these functions are done in a single container in CoreDNS, which is running a process written in Go. The
+different plugins that are enabled replicate (and enhance) the functionality found in `kube-dns`.
```
Comment (neolit123) [inline content/en/blog/_posts/2018-07-10-coredns-ga.md]: [1]

File: content/en/blog/_posts/2018-07-10-coredns-ga.md [hunk 12]
```diff
@@ -0,0 +1,179 @@
...
+
+## What's Next?
+
+CoreDNS is an independent project, and as such is developing many features that are not directly
+related to Kubernetes. However, a number of these will have applications within Kubernetes. For example,
+the upcoming integration with policy engines will allow CoreDNS to make intelligent choices about which endpoint
+to return when a headless service is requested. This could be used to route traffic to a local pod, or
+to a more responsive pod. Many other features are in development, and of course as an open source project, we welcome you to suggest and contribute your own features!
+
+The features and differences described above are just a sampling. There is much more you can do with CoreDNS.
```
Comment (neolit123) [inline content/en/blog/_posts/2018-07-10-coredns-ga.md]: `just a sampling` -> `just an example`?

**Comment by kbarnard10** [comment]
@johnbelamaric Could you make the above edits and change the date to today? Once you do, please confirm and I'll publish this. Thanks!

**Comment by fturib** [comment]
@johnbelamaric : does it make sense to add here an extra paragraph with "if you are using CoreDNS as the DNS service of K8s cluster let us know ... - with an email to coredns.io". ?

File: content/en/blog/_posts/2018-07-10-coredns-ga.md [hunk 13]
```diff
@@ -0,0 +1,179 @@
+---
+layout: blog
+title: "CoreDNS GA for Kubernetes Cluster DNS"
+date: 2018-07-08
+---
+
+**Author**: John Belamaric (Infoblox)
+
```
[the remark this answers is not in the document]
Comment (johnbelamaric) [reply content/en/blog/_posts/2018-07-10-coredns-ga.md]: ok, will do

**Comment by johnbelamaric** [comment]
Ok, I think I have covered all the comments. PTAL

**Pull request merged**